src/util/createServerRootMixin.js

```diff
@@ -40,31 +61,56 @@ function searchOnlyWithDerivedHelpers(helper) {
   });
 }
 
-function defaultCloneComponent(componentInstance) {
+function defaultCloneComponent(componentInstance, { mixins = [] } = {}) {
   const options = {
     serverPrefetch: undefined,
     fetch: undefined,
     _base: undefined,
     name: 'ais-ssr-root-component',
-    // copy over global Vue APIs
-    router: componentInstance.$router,
-    store: componentInstance.$store,
   };
+  if (isVue2) {
+    // copy over global Vue APIs
+    options.router = componentInstance.$router;
+    options.store = componentInstance.$store;
+  }
 
-  const Extended = componentInstance.$vnode
-    ? componentInstance.$vnode.componentOptions.Ctor.extend(options)
-    : (isVue3 ? defineComponent : Vue2.component)(
-        Object.assign({}, componentInstance.$options, options)
-      );
-
-  const app = new Extended({
-    propsData: componentInstance.$options.propsData,
-  });
+  let app;
+
+  if (isVue3) {
+    const appOptions = Object.assign({}, componentInstance.$options, options);
+    appOptions.mixins = [...appOptions.mixins, ...mixins];
+    // Unlike Vue 2, there is no componentInstance.$options.propsData in Vue 3.
+    // The only way to pass the propsData is to spread componentInstance
+    // in the second argument, hoping the rest wouldn't make any side effect.
+    // At this point, we don't even have the definition of the props.
+    // So we cannot pass exactly the propsData only.
+    // FIXME: Maybe we need to get the list of props in `createServerRootMixin`.
+    app = createSSRApp(appOptions, _objectSpread({}, componentInstance));
+    if (componentInstance.$router) {
+      app.use(componentInstance.$router);
+    }
+    if (componentInstance.$store) {
+      app.use(componentInstance.$store);
+    }
+  } else {
+    const Extended = componentInstance.$vnode
+      ? componentInstance.$vnode.componentOptions.Ctor.extend(options)
+      : Vue2.component(Object.assign({}, componentInstance.$options, options));
+
+    app = new Extended({
+      propsData: componentInstance.$options.propsData,
+      mixins: [...mixins],
+    });
+  }
 
   // https://stackoverflow.com/a/48195006/3185307
   app.$slots = componentInstance.$slots;
   app.$root = componentInstance.$root;
-  app.$options.serverPrefetch = [];
+  if (isVue2) {
+    app.$options.serverPrefetch = [];
+  } else {
+    // FIXME: ???
```

Comment:
IIRC, this is to prevent infinite calling of `serverPrefetch`, but this instance variable doesn't exist in vue 3 and without it, it seems okay so far.

Reply:
it's thanks to https://github.com/algolia/vue-instantsearch/pull/1020/files#diff-ac44dbb445b26102c11a145082de41832504ef96bb746a1a05387f759c3d8909L45 in vue 3

Reply:
Ah you're right!